### PR TITLE
zsh: add support for prezto configuration framework

### PR DESCRIPTION
**Pull request by NickHu**
This commit adds support for [prezto](https://github.com/sorin-ionescu/prezto), an alternative to oh-my-zsh. Currently it depends on https://github.com/NixOS/nixpkgs/pull/59134 to some extent (this pull request actually makes it so that `$HOME/.zpreztorc` is sourced).

### CODE REVIEW

**Comment by rycee** [comment]
Thanks for the contribution! I'm a bit worried about the size of this, though. Maintaining this will be a large job. And I'm worried about having two completely separate configuration frameworks in the same HM module.

If it's added then perhaps could have something simpler like
```
prezto = {
  settings = mkOption { type = types.attrsOf types.str; … };
  load = mkOption { type = types.attrsOf types.str; … };
  moduleConfig = mkOption { type = types.attrsOf (types.attrsOf types.str); … };
};
```
so that a configuration could be along the lines of
```
{
  prezto = {
    settings.case-sensitive = true;
    load = {
      pmodule-dirs = "$HOME/.zprezto-contrib";
    };
    module = {
      prompt.pwd-length = "long";
    };
  };
}
```

But I usually excuse myself when it comes to zsh module changes so I'll defer to the opinion of the zsh expert @uvNikita 🙂

**Comment by NickHu** [comment]
The reason for its size is so that everything can be configured from home-manager; I'm not sure that it should require much maintenance.

**Comment by teto** [comment]
Seems like there is some overlap between these ZSH frameworks and nix. There are settings for python env, more plugins, ruby, ssh identities. IMHO, the most value is made of nix is when settings are redundant or interdependant between software. This seems like it can remain selfcontained so why not just copy the configuration files ?

**Comment by NickHu** [comment]
@teto Yes, there is some overlap. The main advantage of these frameworks is that they're able to configure various plugins with minimal setup; nobody has bothered to write a nix-centric zsh plugin manager, so the next best thing is to co-opt an existing one in this way (as home-manager already does with `oh-my-zsh`).

@rycee It's been a year now and I've taken to patching home-manager to include this every time I update, so I would really like some incarnation of this to be merged. I think I was right in saying that it's effectively no-maintenance, so how would you feel about merging now? Perhaps a neat way of splitting it up is to put most of the implementation in its own file (say something like `modules/programs/zsh/prezto.nix`) and using the `import` mechanism in the main zsh nix expression. I doubt the implementation would ever need to be changed, and I'm happy to take on responsibility for updating it if it ever does.

**Comment by dkudriavtsev** [comment]
I am also very interested in seeing this merged. I am currently using NickHu's fork of home-manager, which is understandably behind by many commits

**Comment by NickHu** [comment]
I've moved the prezto stuff to its own file to keep `zsh.nix` a bit slimmer; like I said before, prezto has a very *very* stable interface and this patch has applied cleanly since I wrote it over a year ago. If there are no objections then I'll add myself to the codeowners and merge this in.

File: modules/programs/zsh.nix [hunk 1]
```diff
@@ -360,15 +368,39 @@ in
     })
 
     (mkIf (cfg.profileExtra != "") {
-      home.file."${relToDotDir ".zprofile"}".text = cfg.profileExtra;
+      home.file."${relToDotDir ".zprofile"}".text =
```
Comment (rycee) [inline modules/programs/zsh.nix:371]: Could try to use something like
```nix
(mkIf cfg.prezto.enable {
  programs.zsh.profileExtra = builtins.readFile "${pkgs.zsh-prezto}/runcoms/zprofile";
})
```
maybe
```nix
mkIf cfg.prezto.enable {
  programs.zsh.profileExtra = mkBefore (builtins.readFile "${pkgs.zsh-prezto}/runcoms/zprofile");
}
```
if the prezto code need to be early in the profile.

If it works then it should be possible to remove the `(mkIf (cfg.prezto.enable && cfg.profileExtra == "") {` below.

Comment (rycee) [reply modules/programs/zsh.nix:371]: Actually, perhaps simply put
```
programs.zsh.profileExtra = builtins.readFile "${pkgs.zsh-prezto}/runcoms/zprofile";
```
in the `prezto.nix` file?

File: modules/programs/zsh.nix [hunk 2]
```diff
@@ -395,7 +429,8 @@ in
     {
       home.packages = with pkgs; [ zsh ]
         ++ optional cfg.enableCompletion nix-zsh-completions
-        ++ optional cfg.oh-my-zsh.enable oh-my-zsh;
+        ++ optional cfg.oh-my-zsh.enable oh-my-zsh
+        ++ optional cfg.prezto.enable zsh-prezto;
```
Comment (rycee) [inline modules/programs/zsh.nix:433]: Could move this to the `prezto.nix` file?

File: modules/programs/zsh.nix [hunk 3]
```diff
@@ -378,6 +410,8 @@ in
       '';
     })
 
+    (mkIf cfg.prezto.enable prezto.config)
```
Comment (rycee) [inline modules/programs/zsh.nix:413]: Move to `prezto.nix`?

File: modules/programs/zsh.nix [hunk 4]
```diff
@@ -360,15 +368,39 @@ in
...
 
     (mkIf (cfg.logoutExtra != "") {
-      home.file."${relToDotDir ".zlogout"}".text = cfg.logoutExtra;
+      home.file."${relToDotDir ".zlogout"}".text =
+        optionalString cfg.prezto.enable
+        (builtins.readFile "${pkgs.zsh-prezto}/runcoms/zlogout")
+        + cfg.logoutExtra;
+    })
+
+    (mkIf (cfg.prezto.enable && cfg.logoutExtra == "") {
```
Comment (rycee) [inline modules/programs/zsh.nix:401]: Move to `prezto.nix`?

File: modules/programs/zsh.nix [hunk 5]
```diff
@@ -360,15 +368,39 @@ in
...
+
+    (mkIf (cfg.prezto.enable && cfg.loginExtra == "") {
+      home.file."${relToDotDir ".zlogin"}".source =
+        "${pkgs.zsh-prezto}/runcoms/zlogin";
     })
 
     (mkIf (cfg.logoutExtra != "") {
-      home.file."${relToDotDir ".zlogout"}".text = cfg.logoutExtra;
+      home.file."${relToDotDir ".zlogout"}".text =
+        optionalString cfg.prezto.enable
```
Comment (rycee) [inline modules/programs/zsh.nix:396]: Move to `prezto.nix`?

File: modules/programs/zsh.nix [hunk 6]
```diff
@@ -360,15 +368,39 @@ in
...
 
     (mkIf (cfg.loginExtra != "") {
-      home.file."${relToDotDir ".zlogin"}".text = cfg.loginExtra;
+      home.file."${relToDotDir ".zlogin"}".text =
+        optionalString cfg.prezto.enable
+        (builtins.readFile "${pkgs.zsh-prezto}/runcoms/zlogin")
+        + cfg.loginExtra;
+    })
+
+    (mkIf (cfg.prezto.enable && cfg.loginExtra == "") {
```
Comment (rycee) [inline modules/programs/zsh.nix:389]: Eh, and so on…

File: modules/programs/zsh.nix [hunk 7]
```diff
@@ -170,6 +170,8 @@ let
     };
   };
 
+  prezto = import ./zsh/prezto.nix { inherit lib pkgs relToDotDir cfg; };
```
Comment (rycee) [inline modules/programs/zsh.nix:173]: Can import in `modules.nix` and maintain it more separately from `zsh.nix`.

**Comment by rycee** [comment]
Added a few comments but since I don't use zsh I can't review properly. If you don't agree with the comments then feel free to merge.

**Pull request merged**